--- templates/edittoolbar.nat.tmpl.dist	2007-03-08 09:09:02.000000000 +0100
+++ templates/edittoolbar.nat.tmpl	    2007-08-16 10:49:30.000000000 +0200
@@ -1,5 +1,6 @@
 <!-- template: edittoolbar.nat.tmpl -->
-%TMPL:DEF{"edittoolbar"}%<div class="natEditToolBar">%TMPL:P{"texttools"}%%TMPL:P{"headlinetools"}% %TMPL:P{"paragraphtools"}% %TMPL:P{"objecttools"}% %TMPL:P{"edittools"}%%TMPL:P{"exittools"}%%TMPL:P{"helptools"}%</div>%TMPL:END%
+%TMPL:INCLUDE{"editcaptcha"}%
+%TMPL:DEF{"edittoolbar"}%<div class="natEditToolBar">%TMPL:P{"texttools"}%%TMPL:P{"headlinetools"}% %TMPL:P{"paragraphtools"}% %TMPL:P{"objecttools"}% %TMPL:P{"edittools"}%%TMPL:P{"exittools"}%%TMPL:P{"helptools"}%%TMPL:P{"editcaptcha"}%</div>%TMPL:END%
 %TMPL:DEF{"texttools"}%<span class="natEditTextTools"><!--
 --><a class="natEditBoldButton" accesskey="b" href="javascript:natEditBoldButtonAction();"><!--
 --><img src="%PUBURLPATH%/%TWIKIWEB%/NatSkin/button_bold.png" alt='%MAKETEXT{"Bold text"}%' title='%MAKETEXT{"Bold text"}%'/></a><!--
