*** lib/TWiki/Access.pm      Mon Jul 21 18:35:21 2003
--- lib/TWiki/Access.pm   Mon Jul 21 18:35:29 2003
***************
*** 115,122 ****
      }
  
      my $webPrefix = "";
!     if( $theWebName ne $TWiki::webName && $theWebName ne $TWiki::twikiWebname && 
!         $theWebName ne $TWiki::mainWebname) {
          # Different Web to current one, but we assume read access to twiki and main webs to
          # save frequent loading of these preferences
          $webPrefix = "Web$theWebName.";
--- 115,121 ----
      }
  
      my $webPrefix = "";
!     if( $theWebName ne $TWiki::webName ) {
          # Different Web to current one, but we assume read access to twiki and main webs to
          # save frequent loading of these preferences
          $webPrefix = "Web$theWebName.";
