*** lib/TWiki.pm	Tue Mar 25 22:47:25 2003
--- lib/TWiki.pm	Tue Mar 25 22:47:44 2003
***************
*** 592,598 ****
  
  	}
      }
!     $finalHeaders .= "\r\n";		
      ##writeDebug( "===== Final Headers are:\n$finalHeaders" );
      print $finalHeaders;
  
--- 592,598 ----
  
  	}
      }
!     if ( $finalHeaders) { $finalHeaders .= "\r\n"; } 
      ##writeDebug( "===== Final Headers are:\n$finalHeaders" );
      print $finalHeaders;
  
