*** lib/TWiki/Prefs.pm        Sat Jan  4 17:36:32 2003
--- lib/TWiki/Prefs.pm      Sat Apr 12 10:11:45 2003
***************
*** 200,205 ****
--- 248,255 ----
              @finalPrefsKeys = ();
              my @saveKeys    = @prefsKeys; # quick hack, this stinks
              my @saveValues  = @prefsValues; # ditto
+             @prefsKeys      = ();
+             @prefsValues    = ();
              getPrefsFromTopic( $TWiki::twikiWebname, $TWiki::wikiPrefsTopicname );
              getPrefsFromTopic( $altWebName, $TWiki::webPrefsTopicname );
              @altPrefsKeys   = @prefsKeys; # quick hack, this stinks
