*** lib/TWiki/Store/RcsWrap.pm      2003-01-04 17:37:22.000000000 -0800
--- lib/TWiki/Store/RcsWrap.pm  2003-10-18 23:45:20.000000000 -0700
***************
*** 157,164 ****
--- 157,167 ----
      }
      $self->_saveFile( $self->file(), $text );
      $cmd = $self->{ciDateCmd};
+     $date =~ s/$TWiki::securityFilter//go;
      $cmd =~ s/%DATE%/$date/;
      $cmd =~ s/%USERNAME%/$user/;
+     $file =~ s/$TWiki::securityFilter//go;
+     $rcsFile =~ s/$TWiki::securityFilter//go;
      $cmd =~ s/%FILENAME%/$file $rcsFile/;
      $cmd =~ /(.*)/;
      $cmd = $1;       # safe, so untaint variable
***************
*** 383,390 ****
      my $cmd = $self->{"ciCmd"};
      my $rcsOutput = "";
      $cmd =~ s/%USERNAME%/$userName/;
      $cmd =~ s/%FILENAME%/$file/;
!     $comment = "none" if ( ! $comment );
      $cmd =~ s/%COMMENT%/$comment/;
      $cmd =~ /(.*)/;
      $cmd = $1;       # safe, so untaint variable
--- 386,395 ----
      my $cmd = $self->{"ciCmd"};
      my $rcsOutput = "";
      $cmd =~ s/%USERNAME%/$userName/;
+     $file =~ s/$TWiki::securityFilter//go;
      $cmd =~ s/%FILENAME%/$file/;
!     $comment = "none" unless( $comment );
!     $comment =~ s/[\"\'\`\;]//go;  # security, Codev.NoShellCharacterEscapingInFileAttachComment, MikeSmith
      $cmd =~ s/%COMMENT%/$comment/;
      $cmd =~ /(.*)/;
      $cmd = $1;       # safe, so untaint variable
