--- lib/TWiki.cfg	2004-09-01 02:35:19.000000000 +1000
+++ lib/TWiki.cfg	2005-01-16 12:50:00.000000000 +1100
@@ -367,8 +367,10 @@
 #                   (e.g. 'fsmith') to Wiki name (e.g. 'FredSmith') :
 $userListFilename   = "$dataDir/$mainWebname/$wikiUsersTopicname.txt";
 #                   Map login name to Wiki name, default "1", set to "0" for .htpasswd authenticated sites :
-$doMapUserToWikiName = "1";
+$doMapUserToWikiName = "0";
 #                   %HOMETOPIC% : Name of main topic in a web, default "WebHome" :
+#        Map login name to Wiki name, default "1", set to "0" for .htpasswd authenticated sites :
+$doMapUserToWikiName = "0";
 $mainTopicname      = "WebHome";
 #                   %NOTIFYTOPIC% : Name of topic for email notifications, default "WebNotify" :
 $notifyTopicname  = "WebNotify";
