--- lib/TWiki/Net.pm      Sun Jan  5 01:36:11 2003
+++ lib/TWiki/Net.pm      Wed Jul  9 15:04:16 2003
@@ -53,7 +53,16 @@
     my $base64;
     my $result = '';
     $theUrl = "/" unless( $theUrl );
-    my $req = "GET $theUrl HTTP/1.1\r\n";
+    my $req = "GET $theUrl HTTP/1.0\r\n";
+
+    # RNF 19 Apr 2002 Support for outbound proxies.
+    my $proxyHost = &TWiki::Prefs::getPreferencesValue("PROXYHOST");
+    my $proxyPort = &TWiki::Prefs::getPreferencesValue("PROXYPORT");
+    if($proxyHost && $proxyPort) {
+        $req = "GET http://$theHost$theUrl HTTP/1.0\r\n";
+        $theHost = $proxyHost;
+        $thePort = $proxyPort;
+    }

     # RNF 22 Jan 2002 Support for vhosts and user authentication.
     $req .= "Host: $theHost\r\n";
@@ -66,14 +75,6 @@
         $req .= "Authorization: Basic $base64";
     }

-    # RNF 19 Apr 2002 Support for outbound proxies.
-    my $proxyHost = &TWiki::Prefs::getPreferencesValue("PROXYHOST");
-    my $proxyPort = &TWiki::Prefs::getPreferencesValue("PROXYPORT");
-    if($proxyHost && $proxyPort) {
-        $theHost = $proxyHost;
-        $thePort = $proxyPort;
-    }
-
     $req .= $theHeader if( $theHeader );
     $req .= "\r\n\r\n";
