*** lib/TWiki.pm    14 Dec 2003 07:25:57 -0000      1.252
--- lib/TWiki.pm    14 Dec 2003 08:54:39 -0000
***************
*** 813,818 ****
--- 813,823 ----
      $remoteUser = $1;  # untaint variable

      my $remoteAddr = $ENV{'REMOTE_ADDR'} || "";
+
+     if( $ENV{'REDIRECT_STATUS'} && $ENV{'REDIRECT_STATUS'} eq '401' ) {
+         # bail out if authentication failed
+         $remoteAddr = "";
+     }

      if( ( ! $doRememberRemoteUser ) || ( ! $remoteAddr ) ) {
         # do not remember IP address
----------------------------
*** lib/TWiki/Plugins.pm  23 Sep 2003 19:49:19 -0000      1.26
--- lib/TWiki/Plugins.pm  14 Dec 2003 09:16:12 -0000
***************
*** 210,215 ****
--- 210,220 ----
      @activePluginWebs = ();
      @instPlugins = ();

+     if( $ENV{'REDIRECT_STATUS'} && $ENV{'REDIRECT_STATUS'} eq '401' ) {
+         # bail out if authentication failed
+         return "";
+     }
+
      # Get INSTALLEDPLUGINS and DISABLEDPLUGINS variables
      my $plugin = &TWiki::Prefs::getPreferencesValue( "INSTALLEDPLUGINS" ) || "";
      $plugin =~ s/[\n\t\s\r]+/ /go;
